server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java

```diff
@@ -0,0 +1,270 @@
+            ),
+            new TransportResponseHandler<AddVotingConfigExclusionsResponse>() {
+                @Override
+                public void handleResponse(AddVotingConfigExclusionsResponse response) {
+                    listener.onResponse(null);
+                }
+
+                @Override
+                public void handleException(TransportException exp) {
+                    listener.onFailure(exp);
+                }
+
+                @Override
+                public String executor() {
+                    return ThreadPool.Names.SAME;
+                }
+
+                @Override
+                public AddVotingConfigExclusionsResponse read(StreamInput in) throws IOException {
+                    return new AddVotingConfigExclusionsResponse(in);
+                }
+            }
+        );
+    }
+
+    /**
+     * Transport call to clear voting config exclusion
+     *
+     * @param listener callback for response or failure
+     */
+    public void clearVotingConfigExclusion(ActionListener<Void> listener, boolean waitForRemoval) {
+        final ClearVotingConfigExclusionsRequest clearVotingConfigExclusionsRequest = new ClearVotingConfigExclusionsRequest();
+        clearVotingConfigExclusionsRequest.setWaitForRemoval(waitForRemoval);
+        transportService.sendRequest(
+            transportService.getLocalNode(),
+            ClearVotingConfigExclusionsAction.NAME,
+            clearVotingConfigExclusionsRequest,
+            new TransportResponseHandler<ClearVotingConfigExclusionsResponse>() {
+                @Override
+                public void handleResponse(ClearVotingConfigExclusionsResponse response) {
+                    listener.onResponse(null);
+                }
+
+                @Override
+                public void handleException(TransportException exp) {
+                    listener.onFailure(exp);
+                }
+
+                @Override
+                public String executor() {
+                    return ThreadPool.Names.SAME;
+                }
+
+                @Override
+                public ClearVotingConfigExclusionsResponse read(StreamInput in) throws IOException {
+                    return new ClearVotingConfigExclusionsResponse(in);
+                }
+            }
+        );
+    }
+
+    /**
+     * This method triggers batch of tasks for nodes to be decommissioned using executor {@link NodeRemovalClusterStateTaskExecutor}
+     * Once the tasks are submitted, it waits for an expected cluster state to guarantee
+     * that the expected decommissioned nodes are removed from the cluster
+     *
+     * @param nodesToBeDecommissioned set of the node to be decommissioned
+     * @param reason reason of removal
+     * @param timeout timeout for the request
+     * @param nodesRemovedListener callback for the success or failure
+     */
+    public synchronized void removeDecommissionedNodes(
```

Comment:
Do we plan on removing this?

Reply:
You mean the `synchronised`?

I thought to keep it. As it would be good if only one thread does the predicate matching before hitting the observer. Although we are not gaining much with this as this method is idempotent. But just a minor case where we can get into two observers waiting (if the first one already matched)..trying to prevent that